client/my-sites/stats/stats-chart-tabs/index.jsx

```diff
@@ -163,141 +91,87 @@ class StatModuleChartTabs extends Component {
+				<StatsModulePlaceholder className="is-chart" isLoading={ isActiveTabLoading } />
+				<Chart
+					barClick={ this.props.barClick }
+					data={ this.props.chartData }
+					loading={ isActiveTabLoading }
+				/>
+				<StatTabs
+					data={ this.props.counts }
+					tabs={ this.props.charts }
+					switchTab={ this.props.switchTab }
+					selectedTab={ this.props.chartTab }
+					activeIndex={ this.props.queryDate }
+					activeKey="period"
+				/>
+			</Card>
 		);
 	}
 }
 
 const NO_SITE_STATE = {
 	siteId: null,
-	data: [],
+	counts: [],
+	chartData: [],
 };
 
 const connectComponent = connect(
-	( state, { period: { period }, chartTab, queryDate } ) => {
+	( state, { activeLegend, period: { period }, chartTab, queryDate } ) => {
 		const siteId = getSelectedSiteId( state );
 		if ( ! siteId ) {
 			return NO_SITE_STATE;
 		}
 
+		const counts = getCountRecords( state, siteId, period );
+		const chartData = buildChartData( activeLegend, chartTab, counts, period, queryDate );
+		const loadingTabs = getLoadingTabs( state, siteId, period );
+		const isActiveTabLoading = loadingTabs.includes( chartTab ) && chartData.length === 0;
 		const quantity = 'year' === period ? 10 : 30;
 		const timezoneOffset = getSiteOption( state, siteId, 'gmt_offset' ) || 0;
 		const date = getQueryDate( queryDate, timezoneOffset, period, quantity );
-
-		// If we are on the default Tab, grab visitors too
-		const quickQueryFields = 'views' === chartTab ? 'views,visitors' : chartTab;
-
-		const query = { unit: period, date, quantity };
-		const quickQuery = { ...query, stat_fields: quickQueryFields };
-		const fullQuery = { ...query, stat_fields: 'views,visitors,likes,comments,post_titles' };
-
-		const quickQueryRequesting = isRequestingSiteStatsForQuery(
-			state,
-			siteId,
-			'statsVisits',
-			quickQuery
-		);
-		const fullQueryRequesting = isRequestingSiteStatsForQuery(
-			state,
-			siteId,
-			'statsVisits',
-			fullQuery
-		);
-
-		const quickQueryData = getSiteStatsNormalizedData( state, siteId, 'statsVisits', quickQuery );
-		const fullQueryData = getSiteStatsNormalizedData( state, siteId, 'statsVisits', fullQuery );
-		const data = fullQueryRequesting ? quickQueryData : fullQueryData;
+		const queryKey = `${ date }-${ period }-${ quantity }-${ siteId }`;
+		const query = { chartTab, date, period, quantity, siteId, statFields: QUERY_FIELDS };
 
 		return {
-			data,
-			fullQuery,
-			fullQueryRequesting,
-			isActiveTabLoading: quickQueryRequesting && fullQueryRequesting && ! ( data && data.length ),
-			quickQuery,
-			quickQueryRequesting,
+			chartData,
+			counts,
+			isActiveTabLoading,
+			loadingTabs,
+			query,
+			queryKey,
 			siteId,
 		};
 	},
-	{ recordGoogleEvent },
+	{ recordGoogleEvent, requestChartCounts },
 	null,
 	{
 		areStatePropsEqual: compareProps( {
-			deep: [ 'activeTab', 'fullQuery', 'quickQuery' ],
+			shallow: [ 'activeTab', 'isActiveTabLoading' ],
+			deep: [ 'query', 'loadingTabs' ],
```

Comment:
I think in general the preferred approach tends to be memoization on the selectors (using e.g. `createSelector`), rather than performing deep comparisons. Here it shouldn't be too bad, though, since the objects are pretty much known quantities and not too complex.

In any case, I'll take a look to see if any re-renders are happening and whether we need to make any changes here.

Reply:
I did some checking, and this component does seem to be causing unnecessary re-renders. `chartData`, specifically, is causing re-renders further down the line, in the `Chart` and `ChartBarContainer` components, because the generated data isn't equal by reference even though it's equal by value.

In any case, feel free to leave this for me to look at after your PR, since the problem was already there :)

Reply:
Good catch; I'll leave it alone for this PR. 

Feel free to mark me down as a reviewer for your next pass :)